crates/nu-cli/src/repl.rs

```diff
@@ -149,361 +122,472 @@ pub fn evaluate_repl(
 
     kitty_protocol_healthcheck(engine_state);
 
+    // Setup initial engine_state and stack state
+    let mut previous_engine_state = engine_state.clone();
+    let mut previous_stack = stack.clone();
     loop {
-        let loop_start_time = std::time::Instant::now();
+        // clone these values so that they can be moved by AssertUnwindSafe
+        // If there is a panic within this iteration the last engine_state and stack
+        // will be used
+        let mut current_engine_state = previous_engine_state.clone();
+        let mut current_stack = previous_stack.clone();
+        let temp_file_cloned = temp_file.clone();
+        let mut nu_prompt_cloned = nu_prompt.clone();
+
+        match catch_unwind(AssertUnwindSafe(move || {
+            match loop_iteration(
+                &mut current_engine_state,
+                &mut current_stack,
+                line_editor,
+                &mut nu_prompt_cloned,
+                &temp_file_cloned,
+                use_color,
+                &mut entry_num,
+            ) {
+                // pass the most recent version of the line_editor back
+                Ok((continue_loop, line_editor)) => (
+                    Ok(continue_loop),
+                    current_engine_state,
+                    current_stack,
+                    line_editor,
+                ),
+                Err(e) => (
+                    Err(e),
+                    current_engine_state,
+                    current_stack,
+                    Reedline::create(),
+                ),
+            }
+        })) {
+            Ok((result, es, s, le)) => {
+                // setup state for the next iteration of the repl loop
+                previous_engine_state = es;
+                previous_stack = s;
+                line_editor = le;
+                match result {
+                    Ok(false) => {
+                        break;
+                    }
+                    Err(e) => {
+                        return Err(e);
+                    }
+                    _ => (),
+                }
+            }
+            Err(_) => {
+                // line_editor is lost in the error case so reconstruct a new one
```

Comment:
The `Mutex`s in `EngineState` are behind `Arc`s, so `current_engine_state` and `previous_engine_state` share references to the same `Mutex`s . We should probably check whether any of the `Mutex`s in the `previous_engine_state` are poisoned here using `Mutex::is_poisoned` and then set them to a fresh value if so.

One of the fields behind a `Mutex` is `regex_cache`. Since recreating `regex_cache` needs a dependency (`lru`), I guess it would be best to add a `EngineState::reset_poisoned(&mut self)` function. Or better yet, a `EngineState::recover_from_panic(&mut self)` function.